test/Microsoft.ML.Tests/FeatureImportanceTests.cs

```diff
@@ -0,0 +1,106 @@
+{
+    public sealed class FeatureImportanceTests : BaseTestPredictors
+    {
+        private sealed class ScoreAndContribution
+        {
+            public float X1 { get; set; }
+            public float X2Important { get; set; }
+            public float X3 { get; set; }
+            public float X4Rand { get; set; }
+            public float Score { get; set; }
+            [VectorType(4)]
+            public float[] FeatureContributions { get; set; }
+        }
+
+        public FeatureImportanceTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        /// <summary>
+        /// Features: x1, x2, x3, xRand; y = 10*x1 + 20x2 + 5.5x3 + e, xRand- random, Label y is dependant on xRand.
+        /// Test verifies that feature contribution scores are outputted along with a score for predicted data. 
+        /// </summary>
+        [Fact]
+        public void TestFeatureImportance()
+        {
+            // Setup synthetic dataset.
+            const int numberOfInstances = 1000;
+            var rand = new Random(10);
+            float[] yArray = new float[numberOfInstances],
+                x1Array = new float[numberOfInstances],
+                x2Array = new float[numberOfInstances],
+                x3Array = new float[numberOfInstances],
+                x4RandArray = new float[numberOfInstances];
+
+            for (var i = 0; i < numberOfInstances; i++)
+            {
+                var x1 = rand.Next(1000);
+                x1Array[i] = x1;
+                var x2Important = rand.Next(10000);
+                x2Array[i] = x2Important;
+                var x3 = rand.Next(5000);
+                x3Array[i] = x3;
+                var x4Rand = rand.Next(1000);
+                x4RandArray[i] = x4Rand;
+
+                var noise = rand.Next(50);
+                yArray[i] = (float)(10 * x1 + 20 * x2Important + 5.5 * x3 + noise);
+            }
+
+            // Create data view.
+            var bldr = new ArrayDataViewBuilder(Env);
+            bldr.AddColumn("X1", NumberType.Float, x1Array);
+            bldr.AddColumn("X2Important", NumberType.Float, x2Array);
+            bldr.AddColumn("X3", NumberType.Float, x3Array);
+            bldr.AddColumn("X4Rand", NumberType.Float, x4RandArray);
+            bldr.AddColumn("Label", NumberType.Float, yArray);
+            var srcDV = bldr.GetDataView();
+
+            var pipeline = ML.Transforms.Concatenate("Features", "X1", "X2Important", "X3", "X4Rand")
+                .Append(ML.Transforms.Normalize("Features"));
+            var data = pipeline.Fit(srcDV).Transform(srcDV);
+            var model = ML.Regression.Trainers.OnlineGradientDescent().Fit(data);
+            var args = new FeatureImportanceCalculationTransform.Arguments()
+            {
+                Bottom = 10,
+                Top = 10
+            };
+            var output = FeatureImportanceCalculationTransform.Create(Env, args, data, model.Model, model.FeatureColumn);
```

Comment:
>FeatureImportanceCalculationTransform.Create [](start = 25, length = 44)

Hi @ganik thanks for working on this. Please see #581 and #754, and read.

Central to the idea, as we see in #581, is that we are making the things that *trains* a transform, *is* a transform, and *is* the transformed data, three separate things (`IEstimator`, `ITransformer`, `IDataView`) rather than having them conflated in one big huge thing as we have done previously (e.g., the `IDataLoader` and `IDataTransform` interfaces).

We have had, by now, some dozens of issues and PRs meant for the work we are doing to do the conversion from `IDataTransform` based scheme, to the `IEstimator` and `ITransformer` based scheme, as I think you are aware since you have done some of this conversion yourself. (E.g., your work in #1254 where you successfully converted some of the old timeseries transformers over to estimator land.)

So let's review this idiom here. This idiom where you've created the transform *that is also simultaneously the data* is an old concept that was used in the internal tool, but we do not want people to be exposed to because it is incredibly confusing unless you're somehow expecting it. So: if you want to ship this, we should not expose this to users yet, until we structure it properly.

We might consider staging the work so that we internally ship this scorer, but not announce it in any way.

Reply:
totally agreed. The right way to go is to implement estimator API. This is scheduled for 0.9. This particular change is a staged change that doesn't need to get announced.

---
In reply to: [236428108](https://github.com/dotnet/machinelearning/pull/1677#discussion_r236428108) [](ancestors = 236428108)

Reply:
This is great feedback and a good plan. Getting this in 0.8 with the current design would let us address issues like #1675 without needing to wait for the refactoring to be finished.